models/migrations/v110.go

```diff
@@ -17,16 +20,74 @@ func addBranchProtectionCanPushAndEnableWhitelist(x *xorm.Engine) error {
 		RequiredApprovals        int64 `xorm:"NOT NULL DEFAULT 0"`
 	}
 
+	type Review struct {
+		ID       int64 `xorm:"pk autoincr"`
+		Official bool  `xorm:"NOT NULL DEFAULT false"`
+	}
+
 	sess := x.NewSession()
 	defer sess.Close()
 
-	if err := x.Sync2(new(ProtectedBranch)); err != nil {
+	if err := sess.Sync2(new(ProtectedBranch)); err != nil {
 		return err
 	}
 
-	if _, err := x.Exec("UPDATE `protected_branch` SET `can_push` = `enable_whitelist`"); err != nil {
+	if err := sess.Sync2(new(Review)); err != nil {
+		return err
+	}
+
+	if _, err := sess.Exec("UPDATE `protected_branch` SET `can_push` = `enable_whitelist`"); err != nil {
+		return err
+	}
+	if _, err := sess.Exec("UPDATE `protected_branch` SET `enable_approvals_whitelist` = ? WHERE `required_approvals` > ?", true, 0); err != nil {
 		return err
 	}
-	_, err := x.Exec("UPDATE `protected_branch` SET `enable_approvals_whitelist` = ? WHERE `required_approvals` > ?", true, 0)
-	return err
+
+	// Find latest review of each user in each pull request, and set official field if appropriate
+	reviews := []*models.Review{}
+	if x.Dialect().DBType() == core.MSSQL {
+		if err := x.SQL(`SELECT *, max(review.updated_unix) as review_updated_unix FROM review WHERE (review.type = ? OR review.type = ?)
```

Comment:
IMHO this query won't do what you want from it, and it will choke from the sheer number of records. Since Gitea uses sequential ID numbers, you can try this:
```
select * from review where id in (
    select max(id) from review group by issue_id, reviewer_id
    where issue_id > ? and issue_id <= ? and type in (?, ?)
)
```

You will have exactly one review (the last one) per reviewer per issue.

But you'll have to paginate the PR IDs to avoid loading too many records at once.

Reply:
You may have to use:
```
select * from review where id in (
    select max(id), issue_id, reviewer_id from review group by issue_id, reviewer_id
    where issue_id > ? and issue_id <= ? and type in (?, ?)
)
```
for syntactical reasons but I'm not sure. Anyway, it's equivalent.